src/plugin.js

```diff
@@ -63,31 +63,28 @@ function configure(dataset, options) {
 }
 
 function dispatchEvent(chart, listeners, label) {
-	if (!listeners) {
-		return;
-	}
-
-	var context = label.$context;
-	var groups = label.$groups;
+	const context = label.$context;
+	const groups = label.$groups;
 	var callback;
 
-	if (!listeners[groups._set]) {
+	if (!listeners || !listeners[groups._set] || !listeners[groups._set][groups._key]) {
 		return;
 	}
 
 	callback = listeners[groups._set][groups._key];
-	if (!callback) {
-		return;
-	}
 
-	if (helpers.callback(callback, [context]) === true) {
-		// Users are allowed to tweak the given context by injecting values that can be
-		// used in scriptable options to display labels differently based on the current
-		// event (e.g. highlight an hovered label). That's why we update the label with
-		// the output context and schedule a new chart render by setting it dirty.
-		chart[EXPANDO_KEY]._dirty = true;
-		label.update(context);
+	if (helpers.callback(callback, [context]) !== true) {
+		return false;
```

Comment:
It seems for chart.js v3 it is required to explicitly return false, undefined will not work.

Reply:
This method still return `undefined` at line 67, 75 and 80. What is the difference with this line? 

But according [the code](https://github.com/chartjs/Chart.js/blob/c2dd264fe8f0d8895b5f2c2468e8acde5d0118e0/src/core/core.controller.js#L997) for this event logic in v3, Chart.js only handles `false` explicitly to cancel the call of `afterEvent`. This is the [same as v2.9](https://github.com/chartjs/Chart.js/blob/2.9/src/core/core.controller.js#L1014). So I'm not sure to understand why this change is needed. What's the issue when using the previous code (no explicit return)?

Reply:
The previous conditions where undefined is returned are just to handle cases where there is no handler for that event. The case that is tested within the unit tests is related to and handler explicitly not returning true, checking that the chart is not re-rendered: https://github.com/santam85/chartjs-plugin-datalabels/blob/4016b95e905f2f3c9c395d0f18da73698392bf0e/test/specs/events.spec.js#L399

I didn't check the history of Chart.js logic, I decided not to change the expected behaviour in the unit tests, and to find a way to respect the intended behaviour without altering the test.

Reply:
Oh I see, thanks for pointing the reason of this change. I think this was a "bug" in Chart.js (probably fixed by [this PR](https://github.com/chartjs/Chart.js/pull/8049)), so this may not be needed anymore?

Reply:
Good find, will try by linking latest Chart.js

Reply:
@santam85 did you have a chance to test that these explicit returns are not needed?

Reply:
I did, by linking the latest `master`, I still get one failing test:
```1) should not update label when returning not `true`
     events handlers
     Expected spy render not to have been called.
<Jasmine>
@test/specs/events.spec.js:439:30
     Expected spy opacity not to have been called.
<Jasmine>
@test/specs/events.spec.js:440:33
```

Reply:
Thanks for testing!

However the return of `callback` means, if `true`, that the label should be re-evaluated and redrawn (if needed). It doesn't mean to avoid processing the current event if `false` ([this branch](https://github.com/chartjs/Chart.js/blob/c2dd264fe8f0d8895b5f2c2468e8acde5d0118e0/src/core/core.controller.js#L998)). It's a wrong fix which may cause additional issues for some use cases not unit tested here.

In the failing test, [we explicitly use `hover: false`](https://github.com/chartjs/chartjs-plugin-datalabels/blob/master/test/specs/events.spec.js#L414) to prevent any render by the core lib in case of mouse interactions. I suspect that, similar to [this bug](https://github.com/chartjs/Chart.js/issues/8059), the `hover: false` shorthand is also broken and doesn't effectively disable interactions. That could be the reason of this unwanted `render()` call.

@kurkle do you think of any reason why `hover: false` wouldn't be supported anymore?

Reply:
Don't know a reason, but it does not work: https://codepen.io/kurkle/pen/MWeNQrm